vespalib/src/vespa/vespalib/util/relative_frequency_sketch.h

```diff
@@ -0,0 +1,134 @@
+ * indices into our data structure components.
+ *
+ * Important: this frequency sketch _requires_ a good hash function, i.e. high entropy.
+ * Use `RelativeFrequencySketch` with HasGoodEntropyHash=false (default) if this is not the
+ * case for the type being counted, as it implicitly mixes the hash bits using XXH3.
+ *
+ * Thread safety: as thread safe as a std::vector.
+ *
+ * References:
+ *  [0]: The Count-Min Sketch and its Applications (2003)
+ *  [1]: https://en.wikipedia.org/wiki/Count%E2%80%93min_sketch
+ *  [2]: TinyLFU: A Highly Efficient Cache Admission Policy (2015)
+ *  [3]: https://github.com/ben-manes/caffeine/blob/master/caffeine/
+ *       src/main/java/com/github/benmanes/caffeine/cache/FrequencySketch.java
+ */
+class RawRelativeFrequencySketch {
+    alloc::Alloc _buf;
+    size_t       _samples_since_decay;
+    size_t       _window_size;
+    uint32_t     _block_mask_bits;
+public:
+    explicit RawRelativeFrequencySketch(size_t count);
+    ~RawRelativeFrequencySketch();
+
+    void add_by_hash(uint64_t hash) noexcept;
+    [[nodiscard]] std::weak_ordering estimate_relative_frequency_by_hash(uint64_t lhs_hash, uint64_t rhs_hash) const noexcept;
+
+    // Gets the raw underlying counter value saturated in [0, 15] for a given hash.
+    [[nodiscard]] uint8_t count_min_by_hash(uint64_t hash) const noexcept;
+
+    void div_all_by_2() noexcept;
+
+    [[nodiscard]] size_t window_size() const noexcept { return _window_size; }
+};
+
+/**
+ * Wrapper of RawRelativeFrequencySketch for an arbitrary hashable type.
+ *
+ * Only set HasGoodEntropyHash=true if you know that the underlying hash function is
+ * of good quality. This _excludes_ std::hash<> hashes, especially those for integers,
+ * as the hash function for those is more often than not the identity function.
+ *
+ * See `RawRelativeFrequencySketch` for algorithm details.
+ */
+template <typename T, typename Hash = std::hash<T>, bool HasGoodEntropyHash = false>
+requires requires(Hash h, T t) { noexcept(noexcept(h(t))); }
+class RelativeFrequencySketch {
+    RawRelativeFrequencySketch _impl;
+    Hash                       _hash;
+public:
+    // Initializes a sketch used for estimating frequencies for an underlying cache
+    // (or similar datastructure) that can hold a maximum of `count` entries.
+    explicit RelativeFrequencySketch(size_t count, Hash hash = Hash{})
+        : _impl(count),
+          _hash(hash)
+    {}
+    ~RelativeFrequencySketch() = default;
+private:
+    [[nodiscard]] uint64_t hash_elem(const T& elem) const noexcept {
+        uint64_t hash = _hash(elem);
+        if constexpr (!HasGoodEntropyHash) {
+            hash = xxhash::xxh3_64(hash); // Mix it up!
+        }
+        return hash;
+    }
+public:
+    // Increments the estimated frequency for the given element, identified by its hash.
+    // Frequency is saturated at 15.
+    void add(const T& elem) noexcept {
```

Comment:
consider returning the updated min_count here, to avoid having to re-hash the entry to compare with the worst element in the cache later. This would make the estimate_relative_frequency function redundant

Reply:
I've added a separate `add_and_count` method with these semantics—the existing `add` remains `void`. This is because I noticed the function epilogue gets a chain of `CMOV` instructions when extracting the minimum counter value, and I prefer not to include these unless required.

Both functions are simple forwards to a `template <bool ReturnMinCount> uint8_t add_by_hash_impl` function.

[Godbolt example](https://godbolt.org/z/78scd8vY3) (please casually ignore the ugly bunch of added function arguments that emulate `this` 🙈). Also demonstrates the auto-vectorization of the counter block division. Clang in particular turns this into a 256-bytes-at-a-time beast.